File: 1056-capacity-to-ship-packages-within-d-days/solution.py
class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        left, right = max(weights), sum(weights)

        while left <= right:
            mid = left + (right - left) // 2
            if (self.f(weights, mid) > days):
                left = mid + 1
            elif (self.f(weights, mid) < days):
                right = mid - 1
            elif (self.f(weights, mid) == days):
                right = mid - 1
        return left
    
    def f(self, weights, x):
        days = 0
        i = 0
        while (i < len(weights)):
            cap = x
            while (i < len(weights)):
                if cap < weights[i]:
                    break
                else:
                    cap -= weights[i]
                i += 1
            days += 1
        return days

File: 1056-capacity-to-ship-packages-within-d-days/solution.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
        int left = 0;
        int right = 1;
        for (int w : weights) {
            left = max(left, w);
            right += w;
        }

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (f(weights, mid) <= days) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }
    
    // 定义：当运载能力为 x 时，需要 f(x) 天运完所有货物
    // f(x) 随着 x 的增加单调递减
    int f(vector<int>& weights, int x) {
        int days = 0;
        for (int i = 0; i < weights.size(); ) {
            // 尽可能多装货物
            int cap = x;
            while (i < weights.size()) {
                if (cap < weights[i]) break;
                else cap -= weights[i];
                i++;
            }
            days++;
        }
        return days;
    }
};
